### ISSUE TITLE
This method creates a model with a 100% memory leak loop using model. fit()

### ISSUE THREAD

**Issue by phpYj** [tensorflow/tensorflow#80753]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

2.18

### Custom code

Yes

### OS platform and distribution

ubuntu 2.2 or mac m1

### Mobile device

ubuntu 2.2 or mac m1

### Python version

3.11

### Current behavior?

I have tried various methods, but the memory is definitely leaking, it seems that the release of memory cannot keep up. Through the logs, it can be found that there is periodic memory recycling, but with the increase of time, there is still a clear upward trend

### Standalone code to reproduce the issue

```shell
import gc

import keras
import numpy as np
import psutil
from keras.optimizers import Adam
from keras.layers import Dense, Dropout, Input, LSTM
import tensorflow as tf
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from tensorflow.keras.models import Sequential, load_model
from tensorflow.keras.callbacks import EarlyStopping, ModelCheckpoint
import time
import json


num_samples = 6
num_features = 3
num_classes = 4
epochs = 50
batch_size = 2
identifier = "test_model"
num_iterations = 500  

def build_model(X, num_classes):
    model = Sequential()
    model.add(Input(shape=(X.shape[1], X.shape[2])))
    model.add(LSTM(16, return_sequences=True))
    model.add(LSTM(16))
    model.add(Dropout(0.4))
    model.add(Dense(8, activation='tanh'))
    model.add(Dense(num_classes, activation='softmax'))

    model.compile(loss='categorical_crossentropy', optimizer=Adam(), metrics=['categorical_accuracy'])
    return model


data_X = np.random.rand(num_samples, num_features)
data_Y = np.random.randint(0, num_classes, size=(num_samples, 1))  


data_Y = np.eye(num_classes)[data_Y.flatten()]  
print(type(data_X))

scaler = MinMaxScaler()
data_X_scaled = scaler.fit_transform(data_X)


train_X, test_X, train_Y, test_Y = train_test_split(data_X_scaled, data_Y, train_size=0.6, random_state=42)


train_X = np.expand_dims(train_X, axis=1)
test_X = np.expand_dims(test_X, axis=1)

best_loss = np.inf
best_model_data = None
for iteration in range(num_iterations):
   
    tf.keras.backend.clear_session()
    tf.compat.v1.reset_default_graph()
    model = build_model(train_X, num_classes)
 
    model_name = f"model_{iteration}"
    early_stopping = EarlyStopping(monitor='loss', patience=10, verbose=0, restore_best_weights=True)
    print(f"Iteration {iteration + 1}/{num_iterations}")
    process = psutil.Process()
    mem_info = process.memory_info()
    print(f"start Current memory usage: {mem_info.rss / (1024 * 1024):.2f} MB")  # RSS - Resident Set Size
    try:
        history = model.fit(train_X, train_Y, epochs=epochs, batch_size=batch_size, shuffle=True,
                            validation_data=(test_X, test_Y), verbose=0)
        current_loss = history.history['loss'][-1]
        print(f"Training model: {model.name}")
    
        del model
        tf.keras.backend.clear_session()
        gc.collect()
    except Exception as e:
        print("err:", e)
    finally:
        process = psutil.Process()
        mem_info = process.memory_info()
        print(f"end Current memory usage: {mem_info.rss / (1024 * 1024):.2f} MB")  # RSS - Resident Set Size

print("end！")


if best_model_data:
    model_json = best_model_data["model_architecture"]
    model_weights = json.loads(best_model_data["model_weights"], object_hook=lambda d: np.array(d))
    model = tf.keras.models.model_from_json(model_json)
    model.set_weights(model_weights)
    model.compile(loss='categorical_crossentropy', optimizer=Adam(), metrics=['categorical_accuracy'])
    print("ok")
else:
    print("not found")
```

### Relevant log output

```shell
Iteration 1/500
start Current memory usage: 450.69 MB
Training model: sequential
end Current memory usage: 524.41 MB
Iteration 2/500
start Current memory usage: 524.52 MB
Training model: sequential
end Current memory usage: 564.97 MB
Iteration 3/500
start Current memory usage: 564.98 MB
Training model: sequential
end Current memory usage: 598.00 MB
Iteration 4/500
start Current memory usage: 598.03 MB
Training model: sequential
end Current memory usage: 624.69 MB
Iteration 5/500
start Current memory usage: 624.69 MB
Training model: sequential
end Current memory usage: 653.89 MB
Iteration 6/500
start Current memory usage: 653.91 MB
Training model: sequential
end Current memory usage: 679.45 MB
Iteration 7/500
start Current memory usage: 679.45 MB
Training model: sequential
end Current memory usage: 701.59 MB
Iteration 8/500
start Current memory usage: 701.59 MB
Training model: sequential
end Current memory usage: 726.83 MB
Iteration 9/500
start Current memory usage: 726.84 MB
Training model: sequential
end Current memory usage: 749.56 MB
Iteration 10/500
start Current memory usage: 749.56 MB
Training model: sequential
end Current memory usage: 782.56 MB
Iteration 11/500
start Current memory usage: 782.56 MB
Training model: sequential
end Current memory usage: 805.92 MB
Iteration 12/500
start Current memory usage: 805.92 MB
Training model: sequential
end Current memory usage: 833.17 MB
Iteration 13/500
start Current memory usage: 833.17 MB
Training model: sequential
end Current memory usage: 852.84 MB
Iteration 14/500
start Current memory usage: 852.84 MB
Training model: sequential
end Current memory usage: 875.05 MB
Iteration 15/500
start Current memory usage: 875.06 MB
Training model: sequential
end Current memory usage: 901.56 MB
Iteration 16/500
start Current memory usage: 901.56 MB
Training model: sequential
end Current memory usage: 930.62 MB
Iteration 17/500
start Current memory usage: 705.70 MB
Training model: sequential
end Current memory usage: 762.64 MB
Iteration 18/500
start Current memory usage: 762.70 MB
Training model: sequential
end Current memory usage: 798.06 MB
Iteration 19/500
start Current memory usage: 798.17 MB
Training model: sequential
end Current memory usage: 824.98 MB
Iteration 20/500
start Current memory usage: 824.98 MB
Training model: sequential
end Current memory usage: 850.34 MB
Iteration 21/500
start Current memory usage: 850.42 MB
Training model: sequential
end Current memory usage: 876.81 MB
Iteration 22/500
start Current memory usage: 876.81 MB
Training model: sequential
end Current memory usage: 904.02 MB
Iteration 23/500
start Current memory usage: 904.08 MB
Training model: sequential
end Current memory usage: 929.70 MB
Iteration 24/500
start Current memory usage: 929.73 MB
Training model: sequential
end Current memory usage: 952.33 MB
Iteration 25/500
start Current memory usage: 952.34 MB
Training model: sequential
end Current memory usage: 952.28 MB
Iteration 26/500
start Current memory usage: 952.47 MB
Training model: sequential
end Current memory usage: 980.39 MB
Iteration 27/500
start Current memory usage: 978.78 MB
Training model: sequential
end Current memory usage: 999.02 MB
Iteration 28/500
start Current memory usage: 999.05 MB
Training model: sequential
end Current memory usage: 1023.50 MB
Iteration 29/500
start Current memory usage: 1023.53 MB
Training model: sequential
end Current memory usage: 1047.80 MB
Iteration 30/500
start Current memory usage: 1047.83 MB
Training model: sequential
end Current memory usage: 1068.88 MB
Iteration 31/500
start Current memory usage: 1068.94 MB
Training model: sequential
end Current memory usage: 1095.78 MB
Iteration 32/500
start Current memory usage: 1095.78 MB
Training model: sequential
end Current memory usage: 1119.03 MB
Iteration 33/500
start Current memory usage: 1119.03 MB
Training model: sequential
end Current memory usage: 1039.41 MB
Iteration 34/500
start Current memory usage: 1022.78 MB
Training model: sequential
end Current memory usage: 1040.88 MB
Iteration 35/500
start Current memory usage: 1040.70 MB
Training model: sequential
end Current memory usage: 1054.58 MB
Iteration 36/500
start Current memory usage: 1054.58 MB
Training model: sequential
end Current memory usage: 1076.16 MB
Iteration 37/500
start Current memory usage: 1076.19 MB
Training model: sequential
end Current memory usage: 1097.02 MB
Iteration 38/500
start Current memory usage: 1097.03 MB
Training model: sequential
end Current memory usage: 1113.70 MB
Iteration 39/500
start Current memory usage: 1114.12 MB
Training model: sequential
end Current memory usage: 1140.30 MB
Iteration 40/500
start Current memory usage: 1140.33 MB
Training model: sequential
end Current memory usage: 1163.81 MB
Iteration 41/500
start Current memory usage: 1163.86 MB
Training model: sequential
end Current memory usage: 1195.83 MB
Iteration 42/500
start Current memory usage: 1195.83 MB
Training model: sequential
end Current memory usage: 1221.53 MB
Iteration 43/500
start Current memory usage: 1221.55 MB
Training model: sequential
end Current memory usage: 1231.09 MB
Iteration 44/500
start Current memory usage: 1231.14 MB
Training model: sequential
end Current memory usage: 1245.78 MB
Iteration 45/500
start Current memory usage: 1199.55 MB
Training model: sequential
end Current memory usage: 1221.59 MB
Iteration 46/500
start Current memory usage: 1221.59 MB
Training model: sequential
end Current memory usage: 1249.11 MB
Iteration 47/500
start Current memory usage: 1249.22 MB
Training model: sequential
end Current memory usage: 1275.50 MB
Iteration 48/500
start Current memory usage: 1259.83 MB
Training model: sequential
end Current memory usage: 1290.91 MB
Iteration 49/500
start Current memory usage: 1285.67 MB
Training model: sequential
end Current memory usage: 1296.75 MB
Iteration 50/500
start Current memory usage: 1296.75 MB
Training model: sequential
end Current memory usage: 1306.59 MB
Iteration 51/500
start Current memory usage: 1306.59 MB
Training model: sequential
end Current memory usage: 1287.53 MB
Iteration 52/500
start Current memory usage: 1287.53 MB
Training model: sequential
end Current memory usage: 1297.23 MB
Iteration 53/500
start Current memory usage: 1297.25 MB
Training model: sequential
end Current memory usage: 1285.45 MB
Iteration 54/500
start Current memory usage: 1285.45 MB
Training model: sequential
end Current memory usage: 1290.36 MB
Iteration 55/500
start Current memory usage: 1282.14 MB
Training model: sequential
end Current memory usage: 1302.14 MB
Iteration 56/500
start Current memory usage: 1302.14 MB
Training model: sequential
end Current memory usage: 1287.70 MB
Iteration 57/500
start Current memory usage: 1287.75 MB
Training model: sequential
end Current memory usage: 1282.77 MB
Iteration 58/500
start Current memory usage: 1271.38 MB
Training model: sequential
end Current memory usage: 1232.14 MB
Iteration 59/500
start Current memory usage: 1212.70 MB
Training model: sequential
end Current memory usage: 1201.16 MB
Iteration 60/500
start Current memory usage: 1200.53 MB
Training model: sequential
end Current memory usage: 1169.45 MB
Iteration 61/500
start Current memory usage: 1169.45 MB
Training model: sequential
end Current memory usage: 1209.73 MB
Iteration 62/500
start Current memory usage: 1207.19 MB
Training model: sequential
end Current memory usage: 1226.28 MB
Iteration 63/500
start Current memory usage: 1226.28 MB
Training model: sequential
end Current memory usage: 1231.45 MB
Iteration 64/500
start Current memory usage: 1210.11 MB
Training model: sequential
end Current memory usage: 1176.00 MB
Iteration 65/500
start Current memory usage: 1173.97 MB
Training model: sequential
end Current memory usage: 1201.42 MB
Iteration 66/500
start Current memory usage: 1201.42 MB
Training model: sequential
end Current memory usage: 1223.94 MB
Iteration 67/500
start Current memory usage: 1222.50 MB
Training model: sequential
end Current memory usage: 1229.80 MB
Iteration 68/500
start Current memory usage: 1227.14 MB
Training model: sequential
end Current memory usage: 1219.02 MB
Iteration 69/500
start Current memory usage: 1210.48 MB
Training model: sequential
end Current memory usage: 1247.17 MB
Iteration 70/500
start Current memory usage: 1245.94 MB
Training model: sequential
end Current memory usage: 1259.84 MB
Iteration 71/500
start Current memory usage: 1259.86 MB
Training model: sequential
end Current memory usage: 1286.39 MB
Iteration 72/500
start Current memory usage: 1286.53 MB
Training model: sequential
end Current memory usage: 1316.52 MB
Iteration 73/500
start Current memory usage: 1311.53 MB
Training model: sequential
end Current memory usage: 1338.72 MB
Iteration 74/500
start Current memory usage: 1338.75 MB
Training model: sequential
end Current memory usage: 1348.45 MB
Iteration 75/500
start Current memory usage: 1338.30 MB
Training model: sequential
end Current memory usage: 1354.97 MB
Iteration 76/500
start Current memory usage: 1353.83 MB
Training model: sequential
end Current memory usage: 1385.67 MB
Iteration 77/500
start Current memory usage: 1385.69 MB
Training model: sequential
end Current memory usage: 1408.83 MB
Iteration 78/500
start Current memory usage: 1408.88 MB
Training model: sequential
end Current memory usage: 1430.91 MB
Iteration 79/500
start Current memory usage: 1430.94 MB
Training model: sequential
end Current memory usage: 1443.62 MB
Iteration 80/500
start Current memory usage: 1428.00 MB
Training model: sequential
end Current memory usage: 1436.50 MB
Iteration 81/500
start Current memory usage: 1436.64 MB
Training model: sequential
end Current memory usage: 1454.66 MB
Iteration 82/500
start Current memory usage: 1440.91 MB
Training model: sequential
end Current memory usage: 1461.81 MB
Iteration 83/500
start Current memory usage: 1460.47 MB
Training model: sequential
end Current memory usage: 1481.19 MB
Iteration 84/500
start Current memory usage: 1481.19 MB
Training model: sequential
end Current memory usage: 1477.84 MB
Iteration 85/500
start Current memory usage: 1477.84 MB
Training model: sequential
end Current memory usage: 1493.55 MB
Iteration 86/500
start Current memory usage: 1493.58 MB
Training model: sequential
end Current memory usage: 1509.50 MB
Iteration 87/500
start Current memory usage: 1509.50 MB
Training model: sequential
end Current memory usage: 1543.94 MB
Iteration 88/500
start Current memory usage: 1542.83 MB
Training model: sequential
end Current memory usage: 1516.17 MB
Iteration 89/500
start Current memory usage: 1516.20 MB
Training model: sequential
end Current memory usage: 1470.17 MB
Iteration 90/500
start Current memory usage: 1470.22 MB
Training model: sequential
end Current memory usage: 1443.72 MB
Iteration 91/500
start Current memory usage: 1444.36 MB
Training model: sequential
end Current memory usage: 1486.23 MB
Iteration 92/500
start Current memory usage: 1476.41 MB
Training model: sequential
end Current memory usage: 1524.97 MB
Iteration 93/500
start Current memory usage: 1524.97 MB
Training model: sequential
end Current memory usage: 1534.94 MB
Iteration 94/500
start Current memory usage: 1551.98 MB
Training model: sequential
end Current memory usage: 1853.48 MB
Iteration 95/500
start Current memory usage: 1853.48 MB
Training model: sequential
end Current memory usage: 1790.12 MB
Iteration 96/500
start Current memory usage: 1792.27 MB
Training model: sequential
end Current memory usage: 1883.20 MB
Iteration 97/500
start Current memory usage: 1879.05 MB
Training model: sequential
end Current memory usage: 1759.69 MB
Iteration 98/500
start Current memory usage: 1669.66 MB
Training model: sequential
end Current memory usage: 1596.77 MB
Iteration 99/500
start Current memory usage: 1597.12 MB
Training model: sequential
end Current memory usage: 1568.83 MB
Iteration 100/500
start Current memory usage: 1532.98 MB
Training model: sequential
end Current memory usage: 1516.75 MB
Iteration 101/500
start Current memory usage: 1465.98 MB
Training model: sequential
end Current memory usage: 1486.66 MB
Iteration 102/500
start Current memory usage: 1483.34 MB
Training model: sequential
end Current memory usage: 1523.19 MB
Iteration 103/500
start Current memory usage: 1523.14 MB
Training model: sequential
end Current memory usage: 1532.77 MB
Iteration 104/500
start Current memory usage: 1531.14 MB
Training model: sequential
end Current memory usage: 1561.78 MB
Iteration 105/500
start Current memory usage: 1555.67 MB
Training model: sequential
end Current memory usage: 1586.70 MB
Iteration 106/500
start Current memory usage: 1586.75 MB
Training model: sequential
end Current memory usage: 1608.41 MB
Iteration 107/500
start Current memory usage: 1603.81 MB
Training model: sequential
end Current memory usage: 1629.00 MB
Iteration 108/500
start Current memory usage: 1629.05 MB
Training model: sequential
end Current memory usage: 1609.25 MB
Iteration 109/500
start Current memory usage: 1609.31 MB
Training model: sequential
end Current memory usage: 1630.09 MB
Iteration 110/500
start Current memory usage: 1629.20 MB
Training model: sequential
end Current memory usage: 1638.66 MB
Iteration 111/500
start Current memory usage: 1620.30 MB
Training model: sequential
end Current memory usage: 1642.81 MB
Iteration 112/500
start Current memory usage: 1642.94 MB
Training model: sequential
end Current memory usage: 1659.45 MB
Iteration 113/500
start Current memory usage: 1655.17 MB
Training model: sequential
end Current memory usage: 1687.80 MB
Iteration 114/500
start Current memory usage: 1673.33 MB
Training model: sequential
end Current memory usage: 1705.94 MB
Iteration 115/500
start Current memory usage: 1699.95 MB
Training model: sequential
end Current memory usage: 1708.22 MB
Iteration 116/500
start Current memory usage: 1707.88 MB
Training model: sequential
end Current memory usage: 1648.23 MB
Iteration 117/500
start Current memory usage: 1634.03 MB
Training model: sequential
end Current memory usage: 1670.97 MB
Iteration 118/500
start Current memory usage: 1671.97 MB
Training model: sequential
end Current memory usage: 1649.69 MB
Iteration 119/500
start Current memory usage: 1645.14 MB
Training model: sequential
end Current memory usage: 1698.64 MB
Iteration 120/500
start Current memory usage: 1699.69 MB
Training model: sequential
end Current memory usage: 1737.67 MB
Iteration 121/500
start Current memory usage: 1737.67 MB
Training model: sequential
end Current memory usage: 1738.05 MB
Iteration 122/500
start Current memory usage: 1721.47 MB
Training model: sequential
end Current memory usage: 1730.64 MB
Iteration 123/500
start Current memory usage: 1729.53 MB
Training model: sequential
end Current memory usage: 1766.12 MB
Iteration 124/500
start Current memory usage: 1761.22 MB
Training model: sequential
end Current memory usage: 1796.58 MB
Iteration 125/500
start Current memory usage: 1796.73 MB
Training model: sequential
end Current memory usage: 1709.02 MB
Iteration 126/500
start Current memory usage: 1721.67 MB
Training model: sequential
end Current memory usage: 1771.50 MB
Iteration 127/500
start Current memory usage: 1771.50 MB
Training model: sequential
end Current memory usage: 1777.38 MB
Iteration 128/500
start Current memory usage: 1757.58 MB
Training model: sequential
end Current memory usage: 1806.50 MB
Iteration 129/500
start Current memory usage: 1758.81 MB
Training model: sequential
end Current memory usage: 1812.45 MB
Iteration 130/500
start Current memory usage: 1812.86 MB
Training model: sequential
end Current memory usage: 1811.14 MB
Iteration 131/500
start Current memory usage: 1799.61 MB
Training model: sequential
end Current memory usage: 1835.33 MB
Iteration 132/500
start Current memory usage: 1716.38 MB
Training model: sequential
end Current memory usage: 1759.75 MB
Iteration 133/500
start Current memory usage: 1752.44 MB
Training model: sequential
end Current memory usage: 1818.41 MB
Iteration 134/500
start Current memory usage: 1811.42 MB
Training model: sequential
end Current memory usage: 1853.58 MB
Iteration 135/500
start Current memory usage: 1853.70 MB
Training model: sequential
end Current memory usage: 1858.50 MB
Iteration 136/500
start Current memory usage: 1858.56 MB
Training model: sequential
end Current memory usage: 1874.84 MB
Iteration 137/500
start Current memory usage: 1862.92 MB
Training model: sequential
end Current memory usage: 1768.23 MB
Iteration 138/500
start Current memory usage: 1762.73 MB
Training model: sequential
end Current memory usage: 1843.39 MB
Iteration 139/500
start Current memory usage: 1843.52 MB
Training model: sequential
end Current memory usage: 1885.88 MB
Iteration 140/500
start Current memory usage: 1885.95 MB
Training model: sequential
end Current memory usage: 1924.86 MB
Iteration 141/500
start Current memory usage: 1925.05 MB
Training model: sequential
end Current memory usage: 1946.80 MB
Iteration 142/500
start Current memory usage: 1946.69 MB
Training model: sequential
end Current memory usage: 1977.53 MB
Iteration 143/500
start Current memory usage: 1974.27 MB
Training model: sequential
end Current memory usage: 1995.17 MB
Iteration 144/500
start Current memory usage: 1992.41 MB
Training model: sequential
end Current memory usage: 1984.45 MB
Iteration 145/500
start Current memory usage: 1963.42 MB
Training model: sequential
end Current memory usage: 1947.31 MB
Iteration 146/500
start Current memory usage: 1944.47 MB
Training model: sequential
end Current memory usage: 1996.00 MB
Iteration 147/500
start Current memory usage: 1996.08 MB
Training model: sequential
end Current memory usage: 2008.41 MB
Iteration 148/500
start Current memory usage: 1999.69 MB
Training model: sequential
end Current memory usage: 1951.30 MB
Iteration 149/500
start Current memory usage: 1942.98 MB
Training model: sequential
end Current memory usage: 1992.28 MB
Iteration 150/500
start Current memory usage: 1982.86 MB
Training model: sequential
end Current memory usage: 2008.83 MB
Iteration 151/500
start Current memory usage: 2008.83 MB
Training model: sequential
end Current memory usage: 1946.42 MB
Iteration 152/500
start Current memory usage: 1946.92 MB
Training model: sequential
end Current memory usage: 1992.48 MB
Iteration 153/500
start Current memory usage: 1979.52 MB
Training model: sequential
end Current memory usage: 2035.66 MB
Iteration 154/500
start Current memory usage: 2023.91 MB
Training model: sequential
end Current memory usage: 2030.31 MB
Iteration 155/500
start Current memory usage: 1974.39 MB
Training model: sequential
end Current memory usage: 2029.30 MB
Iteration 156/500
start Current memory usage: 1997.42 MB
Training model: sequential
end Current memory usage: 2000.31 MB
Iteration 157/500
start Current memory usage: 1964.38 MB
Training model: sequential
end Current memory usage: 1979.45 MB
Iteration 158/500
start Current memory usage: 1973.12 MB
Training model: sequential
```

**Comment by phpYj** [comment]
The main issue is the creation method of the model and the code related to the fit loop, which leads to memory leakage. Other codes have little significance

**Comment by phpYj** [comment]
> The main issue is the creation method of the model and the code related to the fit loop, which leads to memory leakage. Other codes have little significance。 I am running on the CPU

**Comment by phpYj** [comment]
I would like to correct that the version number is 2.18, but I accidentally closed the issue

**Comment by phpYj** [comment]
I would like to correct that the version number is 2.18, but I accidentally closed the issue

**Comment by edwardyehuang** [comment]
same issue, and only occur on GPU (not TPU)

**Comment by Husseinshtia1** [comment]
### 🧠 **CollapseCleaner — A Practical Fix for TensorArray + tf.while\_loop Memory Leaks**

Hi @pavelklinik, @tf-maintainers,

We've encountered the exact class of leak described here — involving `tf.TensorArray` within `tf.while_loop`, especially in Python 3.11 and TensorFlow 2.15–2.17. It occurs when the `TensorArray` grows across iterations and **isn't properly released at runtime**, even if `.close()` is used or out-of-scope.

These fall under what we term **“Eclipse Leaks”** — structural memory residues that accumulate silently in long-looped TensorFlow graphs.

---

### ✅ **CollapseCleaner Identifies & Collapses This Leak**

We’ve tested this behavior on:

* `Python 3.11.1 / 3.11.3`
* `TensorFlow 2.16.1 / nightly`
* `Linux Ubuntu 22.04, Docker`
* 100K+ loop iterations with nested `TensorArray.write(i, ...)`

Without cleanup, memory usage climbed continuously.
**With CollapseCleaner**, memory stabilized within 500–1000 steps.




It identifies orphaned graph components (like persistent references to TensorArray op nodes or unfinished threads) and cleans them up safely — **without forcing session reset**.

---

### 📎 Background + Benchmarks

CollapseCleaner is a diagnostic SDK derived from WaveMind’s CollapseEngine.
It handles memory fragmentation, residual tensor graphs, and CUDA context leakage in:

* `tf.while_loop`
* `tf.TensorArray`
* `tf.data` pipelines
* `tf.function` closures

📖 Read our full post:
🔗 [[CollapseCleaner – The Invisible Leak Draining Billions from AI](https://www.linkedin.com/pulse/invisible-leak-draining-billions-from-ai-until-now-hussein-shtia-h20rf/)](https://www.linkedin.com/pulse/invisible-leak-draining-billions-from-ai-until-now-hussein-shtia-h20rf/)

---

💬 We’d be happy to collaborate with the TensorFlow team on deeper integration or help reproduce the Eclipse Leak via benchmark scripts.

**Comment by vitrion** [comment]
> Dear Tensorflow users and devs.
> 
> I experienced CPU memory leaks during the development of my ML ops framework. I spent over 50 hours searching for a solution
> 
> So i will share my research and final solution.
> 
> How I Identified the RAM Leak :
> 
> Initialy i worked under a Tensorflow NVIDIA at https://catalog.ngc.nvidia.com/orgs/nvidia/containers/tensorflow release 24.10 In order to find the limit of my hyperparameter search space, i did some stress test to find the limit of my hardware (RTX 4500 ada generation). And i found a RAM leak during keras.fit() loop.
> 
> Sorry i don't have screenshoot but the RAM curve is the following
> 
> model.fit() (with jit_compile=True) Phase 1 -> compiling training loop -> Sharp drop of RAM (this is the XLA compiling effect) Phase 2 -> Inverse Log curve -> the first fit steps use more ram than last step and i see a stabilizing effect where the RAM leak become linear)
> 
> During my research i expected the following RAM leaks suspect list -> (Nvidia Driver / Cuda, XLA compiler, tf data pipeline, Tensorflow version, keras version, Grappler optimiser config)
> 
> i spend most of the time suspecting XLA, thus i tried many config flags -> thus XLA suspect removed from list i spend a lot of time trying different container config with (Nvidia hub and tensorflow official docker hub), turn out any NVIDIA driver from 535 to 565 and tf 2.16 ,2.17 ,2.8 still have the issue. -> NVIDIA driver / cuda suspect removed from list i converted the model back to keras 2 API -> still have the issue i run the tf data pipeline without keras.fit -> no memory leak after reading many (maybe all) TF RAM leaks issues, i feel agree with [@phpYj](https://github.com/phpYj)
> 
> so i imagined why internal tensorflow dev still having memory leak issue tickers on github since 2.10 and no fix for many users. and i remembered that tcmalloc is a internal google tools. So i decided to try.
> 
> Thus under tensorflow/tensorflow/2.17.0-gpu container i installed TCmalloc
> 
> BASH :
> 
> apt-get install google-perftools
> 
> find /usr -name "libtcmalloc_minimal.so* (for know the path)
> 
> and run my .py script with this env variable LD_PRELOAD:/usr/lib/x86_64-linux-gnu/libtcmalloc_minimal.so.4 (maybe it will be another path for you)
> 
> CPU RAM dataleak removed !!!
> 
> model.fit() (with jit_compile=True) Phase 1 -> Compiling -> Sharp drop of RAM (this is the XLA compiling effect) Phase 2 -> Training :constant RAM usage with no fluctuation
> 
> I tried again without TCmalloc, and i had RAM leaks again I tried again with TCmalloc and i had NO RAM leaks
> 
> Hope my finding will help some of you
> 
> my nvidia smi under the container -> Driver Version: 535.183.01 CUDA Version: 12.3

I'm facing a similar problem. I've been trying various methods for a long time to prevent memory leaks from affecting the training of my models.

I'm using Mealpy to optimize the hyperparameters of a neural network, ensuring that it cleans the TF session both before and after creating the model. I'm using a Differential Evolution optimizer with 100 individuals per population and 10 iterations in a Docker container with shared_memory=8GB and RAM=24GB, where I'm also using two RTX 4080 GPUs running simultaneously using the MirroredStrategy. Every time, at evaluation 300, my container's RAM fills up and the container kills the process, preventing it from continuing the optimization.

Currently, I'm trying the method you mentioned using the Docker image you recommended and LIBTCMALLOC, but I see that the RAM usage continues to increase even though all the processing is being done on the GPUs. What could be happening? Why isn't the RAM being released?

**Comment by Husseinshtia1** [comment]
Thanks for sharing the details.

I would separate this into two different problems:

1. **Allocator behavior / memory fragmentation**
2. **Real object retention across trials**

`tcmalloc` can help with the first one, but it will not fix the second one.

If RAM keeps increasing consistently around the same evaluation number, for example around evaluation ~300, that usually means something is accumulating across trials. It may not be only a TensorFlow allocator issue. It may be that some references are still alive somewhere in the training or optimization loop.

In your setup, the most suspicious areas are:

* creating a new Keras model on every evaluation
* repeated `compile()` / `fit()` calls creating new graphs
* `MirroredStrategy` resources staying alive
* `History` objects from training being retained
* callbacks keeping references to the model, logs, or tensors
* `tf.data` pipelines being recreated per trial
* Mealpy storing population history, fitness history, trial metadata, or objective outputs
* Python references preventing garbage collection even after `clear_session()`

Also, even when the actual training runs on GPU, CPU RAM is still heavily used. TensorFlow still needs host memory for graph objects, Python objects, dataset iterators, host-side tensor metadata, distributed strategy coordination, callbacks, metric histories, and compiled function caches.

So GPU execution does not mean CPU RAM should remain flat.

What I would check first:

* create `MirroredStrategy` only once, outside the objective function
* make sure the objective returns only a plain scalar fitness value
* do not return or store the model, training history, predictions, tensors, callbacks, or datasets
* avoid recreating datasets inside every trial if possible
* disable or minimize Mealpy history/logging if it stores previous trial objects
* test once without XLA / JIT compilation
* test once with a single GPU and no `MirroredStrategy`
* test the same model-build/train loop without Mealpy to see whether the growth comes from TensorFlow itself or from the optimizer framework retaining state

If RAM still grows after that, I would not rely only on `clear_session()`.

For long hyperparameter searches, the most robust pattern is process isolation: run each trial, or every small batch of trials, in a separate subprocess. When the subprocess exits, the OS releases Python memory, TensorFlow state, CUDA runtime state, graph state, allocator state, and hidden references completely.

It is heavier, but for large optimization loops it is often much more stable than trying to keep one Python/TensorFlow process clean for hundreds or thousands of model builds.

So my guess is:

`tcmalloc` helped in the original case because the main issue was allocator behavior or fragmentation. In your case, the growth may be caused by retained objects, repeated strategy/model/graph creation, or Mealpy keeping some trial state alive.

This is exactly the kind of operational waste we are studying at ML Mind : memory growth, retry waste, GPU/inference inefficiency, cost leaks, and hidden failure patterns inside ML/AI workflows.

You can check it here:
https://mlmind.cloud/

And feel free to reach out if you want to compare notes or share a minimal reproducible example:
hussein@mlmind.cloud

**Comment by vitrion** [comment]
I would like to provide an update on how I successfully resolved my RAM memory leak issue.
All deep neural network processing in TensorFlow with a single GPU must be performed within the objective function. In general, for each new evaluation carried out by the Mealpy objective function, a child process must be created, and all TensorFlow operations should be executed inside that process.
In my case, I had to implement a utility function that allows both building and evaluating the neural network, and that can be used in both the parent process (host) and the child process (worker). This function performs the following steps:

1. Import the TensorFlow and Keras libraries.
2. Create a new TensorFlow session.
3. Define the required callback functions.
4. Build the deep neural network architecture.
5. Compile and train the deep neural network.
6. Note: I was unable to make MirroredStrategy work for multi-GPU training.
7. Evaluate the trained neural network.
8. Return the corresponding performance metric, for example, accuracy. In my case, the function returns the following:
    a) A list containing:
        a.1) The hyperparameters used to build and evaluate the model.
        a.2) The training accuracy at the epoch where the validation loss is minimized.
        a.3) The validation accuracy obtained on the validation dataset.
    b) The resulting trained model.
    c) The validation history.

The worker function is intended solely to invoke the utility function within the child process created for this purpose. Note that the arguments of this function include both the Queue instance created in the parent process (host) and the decoded hyperparameters from the objective function. It performs the following steps:

1. Invoke the utility function using the decoded hyperparameters. In this case, I am only interested in receiving the outputs described in (a) and (c). Keep in mind that the utility function also returns the trained model described in (b); however, since TensorFlow models are not serializable in this context, this return value must be ignored using "_".
2. Push into the Queue instance both the list obtained in (a) and the validation history obtained in (c), since both are serializable.

The Mealpy objective function must follow this structure:

1. Decode the chromosome.
2. If the objective function needs to return, for example, a list containing the hyperparameters evaluated by the model along with their corresponding accuracy, as well as the validation history, a multiprocessing.Queue instance must be created so that these results can be serialized and sent back to the parent process.
3. Create an instance of the Process class to spawn a child process, where the target is the worker function and the args include both the Queue instance and the decoded values from step 1.
4. Start the child process using the start() method.
5. Retrieve the results returned by the worker function from the Queue instance created in step 2 using the get() method.
6. Use the join() method of the Process class to ensure that all data sent from the child process has been received by the parent process.
7. Terminate the child process using the terminate() method to ensure that all resources used by the child process are released, including GPU resources (forcing the cleanup of all TensorFlow graphs generated during the evaluation, which was the source of the RAM memory leak).
8. From the data retrieved in step 5, extract the performance metric to be optimized (in this case, accuracy) and return it.

MirroredStrategy cannot be created within a Mealpy objective function.
The environment variable TF_GPU_ALLOCATOR must be set to cuda_malloc_async to prevent TensorFlow from reserving large VRAM blocks using the traditional allocator. Instead, this allows memory to be allocated asynchronously and released dynamically during execution.

**Comment by Kayyuri** [comment]
Hi,
Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

Could you please try to run each training iteration in a separate multiprocessing child process and terminate it afterward, allowing to fully release all RAM and GPU memory?

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by vitrion** [comment]
My comment should still be valid, since the version of TensorFlow I used is 2.20.0
